vacuum-advisor/vacuum-simulator.mdx

```diff
@@ -0,0 +1,182 @@
+  You can check the Estimated Table Bloat panel of Schema Statistics
+  VACUUM/ANALYZE Activity page for the current estimated table bloat to access
+  the actual impact.
+
+
+## Triggered by freeze age
+
+To prevent transaction ID wraparound, Postgres runs autovacuums periodically to
+freeze older rows. We also say this as "freeze old transaction IDs",
+You can learn more about freezing in the
+[What is freezing?](/docs/vacuum-advisor/what-is-freezing) page.
+
+![Triggered by freeze age](vacuum_simulator_freezing.png)
+
+When either the age of the oldest unfrozen transaction ID or the oldest unfrozen
+multixact ID exceeds `autovacuum_freeze_max_age` or
+`autovacuum_multixact_freeze_max_age` respectively, freeze age autovacuums will
+be triggered. These autovacuums are called anti-wraparound autovacuums.
+
+With VACUUM Simulator, you can tweak `autovacuum_freeze_max_age` and
+`autovacuum_multixact_freeze_max_age` in this section.
+`autovacuum_multixact_freeze_max_age` can be only tweaked if multixact ID is
+used. Lowering these configuration settings triggers autovacuums more, and
+increasing these configuration settings will decrease autovacuums triggered by
+freeze age.
+
+### Limitation
+
+* The simulation is using the data of average transactions (or multixacts) per
+  second (tps).
+  If the tps varies a lot depending on the time of the day or day of the week,
+  the simulation may look a lot different than the actual data.
+* Autovacuums triggered by other causes usually try to freeze old rows.
+  The simulation also freezes them, however, there are cases that freezing won't
+  happen when there are nothing can be frozen, or autovacuum is more focused on
+  cleaning up dead rows.
+  This can cause the transaction age to be reset even when it shouldn't be in
+  the simulation, potentially cause the simulation to trigger less autovacuums
+  by freeze age.
+
+
+## Triggered by inserts
+
+Inserts autovacuum is introduced on Postgres 13.
+With an inserts only table, because it won't generate any dead rows, autovacuums
+won't be triggered by dead rows. It can still be triggered by freeze age, but
+depending on the database workload, it can be a long time until it's triggered.
+Inserts autovacuum is triggered to freeze old transaction IDs like freeze age
+autovacuum, but more importantly, it is triggered to update a visibility map.
+Each table has a visibility map to keep track of which pages contain only tuples
+that are known to be visible to all active transactions. For the case like an
+inserts only table, all the pages of the table should be visible here and less
+likely to be not visible moving forward (will be flagged as not visible if there
+is any write operation on that page).
+However, Postgres doesn't know about it until VACUUM runs to set a bit (flag).
+Updating the visibility map will help to spend less time for any subsequent
+VACUUMs as these pages don't need to be VACUUMed and will be skipped. Updating
```

Comment:
```suggestion
Updating the visibility map will improve performance for any subsequent
VACUUMs as all-visible pages don't need to be VACUUMed and will be skipped. Updating
```
We're kind of blurring the all-visible/all-frozen distinction here, but that's probably okay given the level of detail. I'm kind of torn: on the one hand, I think trying to add in that info might make this too complicated, but on the other it feels like this could be misleading otherwise. Thoughts?

Reply:
Uhmmm good point. After re-reading code:

* freezing vacuum: can't skip all-visible pages (but may skip all-frozen pages)
* dead rows and inserts vacuum: can skip all-visible pages

So "all-visible pages don't need to be VACUUMed" is not 100% accurate. We could add "unless it's a freezing (aggressive) vacuum" (I've been avoiding to use "aggressive vacuum" but I think I should be using as it is commonly used) to make it more accurate without introducing the word of "all-frozen".

Alternatively, inserts vacuum likely flag the page as both all-visible and all-frozen, so we can say something like:

> Updating the visibility map helps index-only scans, as they will be able to
avoid reading the table data to check visibility, improving performance.
> During the VACUUM of an append-only table, it is often able to mark pages as all-frozen, in addition to all-visible, and it'll stay that way moving forward too. This will improve performance for any subsequent VACUUMs as all-visible and all-frozen pages don't need to be VACUUMed and will be skipped.

Thoughts?

Reply:
Yeah, that makes sense. Just minor wording tweaks:

>Updating the visibility map helps index-only scans, as they will be able to
>avoid reading table data to check visibility, improving performance.
>During VACUUM of an append-only table, it is often able to mark pages as all-frozen, in addition to all-visible, and it will stay that way (since these rows will never be modified). This will improve performance for any subsequent VACUUMs as pages that are both all-visible and all-frozen don't need to be VACUUMed and will be skipped.

(The most important is rewording that last sentence to avoid ambiguity.)